File: mobile-app/src/utils/hashEngine.ts
import CryptoJS from 'crypto-js';

/**
 * Generates a SHA-256 integrity hash for a transaction.
 * Canonical format: weight|gps_lat|gps_lng|timestamp_utc|agent_id
 * GPS values fixed to 6 decimal places to avoid floating-point ambiguity.
 */
export function generateTransactionHash(params: {
  weight: number;
  gpsLat: number;
  gpsLng: number;
  timestampUtc: string;
  agentId: string;
}): string {
  const canonical = [
    params.weight.toString(),
    params.gpsLat.toFixed(6),
    params.gpsLng.toFixed(6),
    params.timestampUtc,
    params.agentId,
  ].join('|');

  return CryptoJS.SHA256(canonical).toString(CryptoJS.enc.Hex);
}

export function verifyTransactionHash(
  hash: string,
  params: Parameters<typeof generateTransactionHash>[0],
): boolean {
  return generateTransactionHash(params) === hash;
}

File: mobile-app/lib/core/crypto/hash_engine.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

/// Canonical format: weight|gps_lat|gps_lng|timestamp_utc|agent_id
/// GPS values fixed to 6 decimal places to prevent floating-point ambiguity.
String generateTransactionHash({
  required double weight,
  required double gpsLat,
  required double gpsLng,
  required String timestampUtc,
  required String agentId,
}) {
  final canonical = [
    weight.toString(),
    gpsLat.toStringAsFixed(6),
    gpsLng.toStringAsFixed(6),
    timestampUtc,
    agentId,
  ].join('|');

  final bytes = utf8.encode(canonical);
  return sha256.convert(bytes).toString();
}

bool verifyTransactionHash({
  required String hash,
  required double weight,
  required double gpsLat,
  required double gpsLng,
  required String timestampUtc,
  required String agentId,
}) {
  return generateTransactionHash(
        weight: weight,
        gpsLat: gpsLat,
        gpsLng: gpsLng,
        timestampUtc: timestampUtc,
        agentId: agentId,
      ) ==
      hash;
}
